### ISSUE TITLE
Make Fibers inherit Sequence and be iterable

### ISSUE THREAD

**Issue by gsmaverick** [munificent/wren#155]
Currently iterating over Fibers (especially when using them to represent streams, generators, etc) is quite verbose and the resulting code doesn't read naturally.

### Ideal Use Case

```dart
var counter = new Fiber {
  var count = 0
  while (true) {
    Fiber.yield(count)
    count = count + 1 
  }
}

for (i in counter) {
  IO.print(i)
}
```

The one outstanding question is whether this iterator invokes the `call` or `try` method on the Fiber being iterated. My money would be on `call` since it's the more straightforward way of implementing this. Thoughts?

**Comment by munificent** [comment]
Yeah, I agree `call` is what you want. This has been something I've wanted to do for a long time.

One thing you'll find that's tricky—which is why I haven't gotten around to this yet—is figuring out how to handle the bit of code the fiber executes after the last `yield`. Consider:

```dart
var fiber = new Fiber {
  for (i in 1..3) Fiber.yield(i)
}

while (!fiber.isDone) {
  IO.print(fiber.call)
}
```

You'd think this prints `1`, `2`, `3`, but it actually prints `1`, `2`, `3`, `null`. The last null is because after yielding `3`, the fiber isn't done yet. It still has a tiny bit of code in can execute. There could actually be lots of code after the last yield:

```dart
var fiber = new Fiber {
  for (i in 1..3) Fiber.yield(i)
  IO.print("Whoa more stuff here")
}
```

That means if you treat fibers as sequences that you iterate over until they're done, you'll always get this lame extra value. I haven't spent the time to figure out a good solution for that yet. If you'd like to throw some brainpower at it, I would definitely like to hear what you can come up with.

In my last language, Magpie, I did make fibers extend the base iterable class because I think this is super powerful. It lets you call `map`, and `where` and stuff directly on fibers. Fun! We just have to figure out how to handle that bit of dangling code at the end.

**Comment by gsmaverick** [comment]
How do you feel about having a primitive value to represent that a Fiber has no more values to yield? I have personally found the dangling `null` value you get from calling yield a little bit strange and I feel like it would make for nicer code to have some "value" represent the state of a Fiber that has yielded all of its values.

This is roughly what Python does in its generator implementation; when you call `next()` on a generator that has run out of values it throws a [StopIteration exception](https://docs.python.org/3/library/exceptions.html#StopIteration) to signal this.

**Comment by munificent** [comment]
> How do you feel about having a primitive value to represent that a Fiber has no more values to yield?

I've coincidentally put [quite a bit of thought into just that](http://journal.stuffwithstuff.com/2013/04/17/well-done/). I think it was a cool idea for Magpie, but I'm trying to keep Wren more minimal so I'd like to see if we can live without it.

Either way, it wouldn't help here. The problem is that if you implement `iterate()` on `Fiber`, you don't know if you will have another value until after you try to run the fiber again and see if it completes before yielding.

Actually, maybe there's a more fundamental problem in that a fiber isn't an itera-*ble*, it's effectively an itera-*tor*. In Wren and other languages, there's a separation between the sequence being iterated, and the cursor that represents where we currently are. In Wren the latter is the value passed to and returned from `iterate()`. In other languages it's the `Iterator` or `Enumerator` type instead of `Iterable` or `Enumerable`. When you iterate over a sequence, the cursor changes but the sequence itself does not.

But "iterating" over a fiber is implicitly mutating it. You're running the fiber, which changes its internal state. After it's been iterated once, it can never be iterated again.

Hmm, for kicks I tried actually slapping something together. Check this out:

```dart
class Fiber is Sequence {
  iterate(iterator) {
    var value = call
    if (isDone) return false
    return [value]
  }

  iteratorValue(iterator) { iterator[0] }
}
```

If you put this in the core lib and get it wired up, I believe it may do the right thing. At least it seems to work. The interesting question is does it run the fiber when the user expects? Are there corner cases where this falls down?

**Comment by zeckalpha** [comment]
#28 would affect this.

**Comment by zeckalpha** [comment]
But maybe that is okay? It would force the Fiber to terminate though, and maybe map and where on Fibers should be lazy?

**Comment by munificent** [comment]
Yeah, that's a good point. It would be nice if they were lazy. It would be cool if mapping a fiber gave you a new fiber in return.

The question then is how much weight does it add to the core library, and is it worth it? Maybe for now the best solution is to sit on this until we build up a corpus of real-world code and see if we find ourselves wanting to iterate over fibers in practice.

**Comment by gsmaverick** [comment]
I originally opened this issue when I started working on [wren-test](https://github.com/gsmaverick/wren-test) and my initial implementation seemed to make this issue much more important as a way to clean up some awkward code I had written.

After working with Fibers a bit and evolving all the moving pieces I ended up with some that's simple and doesn't require any boilerplate code. I've simplified the code from its [original source](https://github.com/gsmaverick/wren-test/blob/master/src/runnable.wren#L65) to keep the focus on the issue in question:

```scala
run() {
  _expectations = []

  while (!_fiber.isDone) {
    var result = _fiber.try()

    // Ignore any values that were yielded that weren't an Expectation.
    // Note: When a fiber is finished the last `yield` invocation returns
    // `null` so it will not be added to the array.
    if (result is Expectation) {
      _expectations.add(result)
    }
  }

  return _expectations
}
```

IMO an iterable Fiber wouldn't make this solution any cleaner; in fact I think it might even be detrimental to hide the implementation detail of how the Fiber actually works.
